### PR TITLE
split.data.table minor changes

### LINKED ISSUE #2950: split.data.table is slow? (O(n))
Was trying to create some test data for #2582 and noticed that `split.data.table` does not scale very successfully:

```
library(data.table)
library(microbenchmark)
set.seed(2034909)
sizes = sort(c(x <- 10^(1:6), 5*x))
times = sapply(sizes, function(NN) {
  DT = data.table(
    ID = rep(1:NN, sample(1:3, NN, TRUE)),
    key = 'ID'
  )
  DT[ , V1 := rnorm(.N)]
  t0 = get_nanotime()
  split(DT, by = 'ID')
  get_nanotime() - t0
})

log_n = log(sizes)
log_t = log(times)
png('~/Desktop/split_timing.png')
par(mar = c(5.1, 8.1, 4.1, 2.1))
plot(log(sizes), log(times), 
     xaxt = 'n', yaxt = 'n',
     xlab = 'Resulting Elements', ylab = '',
     main = 'Scaling of split.data.table')
x = 10^(1:7)
axis(side = 1L, at = log(x),
     labels = prettyNum(as.integer(x), big.mark = ','))
yrng = range(ceiling(log10(times) - 9))
y = 10^(yrng[1L]:yrng[2L])
axis(side = 2L, at = log(y) + 9*log(10), las = 1L,
     labels = sprintf('%02d:%02d:%05.2f', 
                      (y %/% 60) %/% 60, (y %/% 60) %% 60, y %% 60))
mtext(side = 2L, 'Time to Split (H:M:S)', line = 6L)
reg = lm(log(times) ~ log(sizes))
abline(reg, lty = 2L)
pu = par('usr')
text(pu[1] + .2*diff(pu[1:2]),
     pu[3] + .8*diff(pu[3:4]),
     labels = bquote(log(t) == .(sprintf('%.1f', coef(reg)['(Intercept)'])) + 
                       .(sprintf('%.1f', coef(reg)['log(sizes)'])) %*% log(n)))
dev.off()
```

split_timing

Basically `time ~ 200,000 * n`, so it's linear but the slope is huge.

Are we using an inefficient algorithm?

### PR DESCRIPTION
**Pull request by jangorecki**
related to #2950 but not resolving it

### CODE REVIEW

File: R/data.table.R [hunk 1]
```diff
@@ -2360,8 +2360,8 @@ split.data.table <- function(x, f, drop = FALSE, by, sorted = FALSE, keep.by = T
     join = TRUE
   }
   dtq[["j"]] = substitute(
-    list(.ll.tech.split=list(.expr)),
-    list(.expr = if (join) quote(if(.N == 0L) .SD[0L] else .SD) else as.name(".SD")) # simplify when `nomatch` accept NULL #857 ?
+    list(.ll.tech.split=list(.expr), .ll.tech.split.names=paste(lapply(.BY, as.character), collapse=".")),
```
Comment (MichaelChirico) [inline R/data.table.R]: no risk of collision here? `by = c('x', 'y')` when the table has `x.y` as a column?

Comment (jangorecki) [reply R/data.table.R]: the risk is pretty low, requires columns to be named `.ll.tech.split` or `.ll.tech.split.names`

**Comment by MichaelChirico** [comment]
@jangorecki committed the change i suggested in #2950 comments

**Comment by jangorecki** [comment]
CI failure is because of the test related to #1908
```
# test if split preallocate columns in results #1908
dt = data.table(x=rexp(100),y=rep(LETTERS[1:10], 10))
dtL = split(dt, by = "y")
test(1639.140, dim(dtL[[1]][, x2 := -x]), c(10L,3L))
test(1639.141, all(sapply(dtL, truelength) > 1000))
```

**Comment by MichaelChirico** [comment]
didn't expect such a test... will leave it in for now while we hash out how to proceed

**Comment by jangorecki** [comment]
@MichaelChirico could you measure what is the speed up after reverting your change to `alloc.col` that causing a failure? maybe it is worth to merge it already as it was before that change.

**Comment by MichaelChirico** [comment]
@jangorecki 

here's the benchmark again

split_timing

**Comment by MichaelChirico** [comment]
Thinking now, if we have `split` result in `1,000,000` `data.table`s with `n=1024` cols to `alloc`, that's over 1B column pointers allocated. At 4 bytes/pointer thats >4Gb, vs. if we minimally-allocate in this case 1e6*2*4=8 Mb for pointers.

If we're not willing to break that test, I don't think there's much to be done... just a heads up in the documentation about this aspect? Or maybe a compromise when `length(split(DT)) > N` for some `N`?

PS a side note, should we edit `alloc.col`->`setalloccol` internally?

**Comment by jangorecki** [comment]
@MichaelChirico I reverted your commit about alloc.col, that piece of code has already changed as uses new `C_unlock` and `setDT`. PR as of now will not reduce the overhead described in linked issue, but I want to merge it anyway as it makes some internals cleaner, and we don't have good idea now how to address mentioned overhead.

**Comment by MichaelChirico** [comment]
Basically the issue comes down to: it may take a while to request memory for a massive amount of pointers.

would it be faster (and possible) to request all the pointers at once rather than table by table? is it something that could be parallelized?

those are the only new thoughts that come to mind

**Comment by jangorecki** [comment]
Likely to give speed up, for now I would like to close this long standing PR :)

**Pull request merged**